firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/Firestore.kt

```diff
@@ -163,3 +161,35 @@ class FirebaseFirestoreKtxRegistrar : ComponentRegistrar {
     override fun getComponents(): List<Component<*>> =
             listOf(LibraryVersionComponent.create(LIBRARY_NAME, BuildConfig.VERSION_NAME))
 }
+
+/**
+ * Attach a snapshotListener to a DocumentReference and use it as a coroutine flow
+ */
+fun DocumentReference.toFlow() = callbackFlow {
+    val listener = addSnapshotListener { value, error ->
+        if (value != null && value.exists()) {
+            offer(value)
+        } else if (error != null) {
+            Logger.warn("DocumentReference:flow", error.message)
```

Comment:
It's not clear to me what the best course of action here would be:
1. i.e. just logging the error is not general enough, as developers may want to react to them and don't want them to be swallowed
2. otoh `close()`ing the flow is not consistent with current Firestore's behavior, as it looks like listeners are not "de-registered" upon exceptions

cc @schmidt-sebastian any thoughts?

Reply:
If the listen stream encounters an error we do currently close the stream. For example, if you have data-dependent rules in effect on the server like a group membership test and a write is performed that removes you from the group, this error is the one that will tell you that permission is denied and that should terminate the stream.

The flow should have the same behavior so this should propagate the error, not swallow it.

Reply:
Thanks Gil,

@michgauz then we definitely want to call `close(error)` here.

Also, as @samtstern mentioned, we will have to have an internal API review for this before it can be merged.

Reply:
Thanks for your review.
I've just taken into account your feedbacks and close the channel when receiving an error 👍